0_From_Code_First_To_First_Principles/1_code_first/assignment/src/controllers/userController.ts

```diff
@@ -0,0 +1,191 @@
+import "reflect-metadata";
+import { AppDataSource } from "../data-source";
+import { User } from "../models/User";
+import { RequestResult } from "../models/RequestResult";
+
+export async function createUser(userJson: any) {
+    try {
+        const user = new User();
+        user.email = userJson.email;
+        user.username = userJson.username
+        user.firstName = userJson.firstName;
+        user.lastName = userJson.lastName;
+        user.password = randomString(10);
+
+        const userValidationFailure = validateUser(user);
+        if (userValidationFailure != null)
+        {
+            return userValidationFailure;
+        }
+
+        const usernameValidationFailure = validateUsername(user.username);
+        if (await usernameValidationFailure != null)
+        {
+            return usernameValidationFailure;
+        }
+
+        const emailValidationFailure = validateEmail(user.email);
+        if (await emailValidationFailure != null)
+        {
+            return emailValidationFailure;
+        }
+        
+        const userRepository = AppDataSource.getRepository(User);
+        await userRepository.save(user);
+        const result = new RequestResult(201, true, undefined, getUserDataJson(user));
+        return result;
+    }
+    catch (error) {
+        console.log(error);
+        const result = new RequestResult(500, false, "ServerError", undefined);
+        return result;
+    }
+}
+
+export async function updateUser(userId: number, userJson: any) {
+    try {
+        const userRepository = AppDataSource.getRepository(User);
+        const user = await userRepository.findOneBy({
+            id: userId,
+        })
+
+        if (user == null)
+        {
+            const userNotFoundResult = getUserNotFoundResult();
+            return userNotFoundResult;
+        }
+
+        const tempUser : User = new User();
+        tempUser.id = user.id,
+        tempUser.email = userJson.email,
+        tempUser.username = userJson.username,
+        tempUser.firstName = userJson.firstName,
+        tempUser.lastName = userJson.lastName
+        tempUser.password = user.password
```

Comment:
it looks like there is an assumption that the user will not update the password?

Reply:
Thanks for the feedback!  I was unsure about this myself.  In the specs for this assignment, it does not list the password as an updatable field.  I thought it may have been left out by accident, or perhaps the password is intended to be updated separately.
Edit User: Accepts a POST to /users/edit/:userId { email, username, firstName, lastName }
https://www.essentialist.dev/products/the-software-essentialist/categories/2153251446/posts/2168948146